enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,667 @@
+		option.WithCredentialsFile(credPath),
+		option.WithEndpoint(fmt.Sprintf("https://%s-cloudresourcemanager.googleapis.com", location)),
+	}
+	client, err := rscmgr.NewTagBindingsRESTClient(ctx, opts...)
+	if err != nil {
+		log.Errorf("failed to create client for tag binding operation: %v", err)
+		return err
+	}
+	defer client.Close()
+
+	tagBindingReq := &rscmgrpb.CreateTagBindingRequest{
+		TagBinding: &rscmgrpb.TagBinding{
+			Parent: fmt.Sprintf("//compute.googleapis.com/projects/%s/zones/%s/instances/%s",
+				project, location, instance),
+			TagValueNamespacedName: tagValueNamespacedName,
+		},
+	}
+
+	opts := []gax.CallOption{
+		gax.WithRetry(func() gax.Retryer {
+			return gax.OnHTTPCodes(gax.Backoff{
+				Initial:    90 * time.Second,
+				Max:        5 * time.Minute,
+				Multiplier: 2,
+			},
+				http.StatusTooManyRequests)
+		}),
+	}
+	result, err := client.CreateTagBinding(ctx, tagBindingReq, opts...)
+	if err != nil {
+		e, ok := err.(*apierror.APIError)
+		if ok && reflect.Indirect(
+		reflect.ValueOf(e).Elem().FieldByName(googleAPIHTTPErrFieldName),
+		).FieldByName(googleAPIHTTPCodeFieldName).Int() == http.StatusConflict {
+			log.Printf("tag binding %s already exist", tag)
+		} else {
+			log.Errorf("tag binding request failed: %v", err)
+			return err
+		}
+	}
+
+	resp, err := result.Wait(ctx)
+	if err != nil {
+		log.Errorf("wait on tag binding operation failed: %v", err)
+		return err
+	}
+
+	log.Printf("tag binding successful: %s: %s", resp.GetName(), resp.String())
+}
+```
+
+Adding labels and tags to GCP resources is controlled using FeatureGate `GCPLabelsTags` and should
+be enabled to activate processing of user defined labels and tags.
+```\h*go\K
+func() {
+	select {
+	case <-featureGateAccessor.InitialFeatureGatesObserved():
+		featureGates, err := featureGateAccessor.CurrentFeatureGates()
+	case <-time.After(1 * time.Minute):
+		log.Error(nil, "timed out waiting for FeatureGate detection")
+		return nil, fmt.Errorf("timed out waiting for FeatureGate detection")
+	}
+
+	gcpLabelsTagsEnabled := featureGates.Enabled(configv1.FeatureGateGCPLabelsTags)
+
+	if gcpLabelsTagsEnabled {
+		// add user defined labels and tags
+	}
+}
+```
+
+#### Caveats
+1. Updating or removing resource labels/tags added by OpenShift using an external interface,
+   may or may not be reconciled by the operator managing the resource.
+2. Updating labels/tags of individual resources is not supported and any label or tag present
+   in `.status.platformStatus.gcp.resourceLabels` or `.status.platformStatus.gcp.resourceTags`
+   respectively of `infrastructure.config.openshift.io/v1` resource will result in adding 
+   labels and tags to all OpenShift managed GCP resources.
+3. Limitations for TechPreview
+   - Worker compute machines created by `machine-api-provider-gcp` controller will not be tagged,
+     controller will be replaced with out-of-tree cloud provider for GCP, and requires tagging
+     support to be added in upstream.
```

Comment:
This isn't quite right, and we don't have timelines for the replacement, so it needs to be added first upstream, then downstream
```suggestion
   - Worker compute machines created by `machine-api-provider-gcp` controller will not be tagged,
     controller will be replaced with cluster-api-provider-gcp, and requires tagging
     support to be added in upstream.
```